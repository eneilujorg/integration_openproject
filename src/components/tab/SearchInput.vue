<template>
	<div id="searchBar">
		<Multiselect
			class="searchInput"
			:placeholder="placeholder"
			:options="searchResults"
			:user-select="true"
			label="displayName"
			track-by="id"
			:internal-search="false"
			open-direction="below"
			:loading="isStateLoading"
			:preselect-first="true"
			:preserve-search="true"
			@search-change="makeSearchRequest"
			@change="linkWorkPackageToFile">
			<template #option="{option}">
				<WorkPackage
					:key="option.id"
					:workpackage="option" />
			</template>
			<template #noOptions>
				{{ translate('Start typing to search') }}
			</template>
		</Multiselect>
		<div v-if="state !== 'ok'"
			class="stateMsg text-center">
			{{ stateMessages }}
		</div>
	</div>
</template>

<script>
import axios from '@nextcloud/axios'
import { generateUrl } from '@nextcloud/router'
import { translate as t } from '@nextcloud/l10n'
import Multiselect from '@nextcloud/vue/dist/Components/Multiselect'
import WorkPackage from './WorkPackage'
import { showError } from '@nextcloud/dialogs'

const STATE_OK = 'ok'
const STATE_ERROR = 'error'
const STATE_NO_TOKEN = 'no-token'
const STATE_LOADING = 'loading'
const SEARCH_CHAR_LIMIT = 3

export default {
	name: 'SearchInput',
	components: {
		Multiselect,
		WorkPackage,
	},
	props: {
		fileInfo: {
			type: Object,
			required: true,
		},
	},
	data: () => ({
		state: STATE_OK,
		searchResults: [],
		selectedId: [],
	}),
	computed: {
		isStateOk() {
			return this.state === STATE_OK
		},
		isStateLoading() {
			return this.state === STATE_LOADING
		},
		placeholder() {
			return this.translate('Search for a work package to create a relation')
		},
		stateMessages() {
			if (this.state === STATE_NO_TOKEN) {
				return this.translate('No OpenProject account connected')
			} else if (this.state === STATE_ERROR) {
				return this.translate('Error connecting to OpenProject')
			}
			return ''
		},
	},
	methods: {
		resetState() {
			this.searchResults = []
			this.state = STATE_OK
		},
		translate(key) {
			return t('integration_openproject', key)
		},
		checkForErrorCode(statusCode) {
			if (statusCode === 200) return
			if (statusCode === 401) {
				this.state = STATE_NO_TOKEN
			} else {
				this.state = STATE_ERROR
			}
		},
		replaceHrefToGetId(href) {
			// this is a helper method replaces the string like this "/api/v3/types/3" to get id
			return href
				? href.replace(/.*\//, '')
				: null
		},
		async linkWorkPackageToFile(selectedOption) {
			const req = {
				values: {
					workpackageId: selectedOption.id,
					fileId: this.fileInfo.id,
				},
			}
			const url = generateUrl('/apps/integration_openproject/work-packages')

			try {
				await axios.post(url, req)
				this.$emit('saved', selectedOption)
				this.selectedId.push({
					id: selectedOption.id,
				})
			} catch (e) {
				showError(
					this.translate('Failed to link file to work-package')
				)
			}
		},
		async makeSearchRequest(search) {
			if (search.length <= SEARCH_CHAR_LIMIT) {
				this.resetState()
				return
			}
			this.state = STATE_LOADING
			const url = generateUrl('/apps/integration_openproject/work-packages')
			const req = {}
			req.params = {
				searchQuery: search,
			}
			let response
			try {
				response = await axios.get(url, req)
			} catch (e) {
				response = e.response
			}
			this.checkForErrorCode(response.status)
			if (response.status === 200) await this.processWorkPackages(response.data)
			if (this.isStateLoading) this.state = STATE_OK
		},
		async processWorkPackages(workPackages) {
			for (const workPackage of workPackages) {
				const statusId = this.replaceHrefToGetId(workPackage._links.status.href)
				const typeId = this.replaceHrefToGetId(workPackage._links.type.href)
				const userId = this.replaceHrefToGetId(workPackage._links.assignee.href)
				const userName = workPackage._links.assignee.title
				const avatarUrl = generateUrl('/apps/integration_openproject/avatar?')
					+ encodeURIComponent('userId')
					+ '=' + userId
					+ '&' + encodeURIComponent('userName')
					+ '=' + userName
				const statusColor = await this.getWorkPackageColorAttributes('/apps/integration_openproject/statuses/', statusId)
				const typeColor = await this.getWorkPackageColorAttributes('/apps/integration_openproject/types/', typeId)
				const selectedIdFound = this.selectedId.some(el => el.id === workPackage.id)
				const workpackageIdFound = this.searchResults.some(el => el.id === workPackage.id)
				if (!workpackageIdFound && !selectedIdFound) {
					this.searchResults.push({
						id: workPackage.id,
						subject: workPackage.subject,
						project: workPackage._links.project.title,
						statusTitle: workPackage._links.status.title,
						typeTitle: workPackage._links.type.title,
						assignee: userName,
						statusCol: statusColor,
						typeCol: typeColor,
						picture: avatarUrl,
					})
				}
			}
		},
		async getWorkPackageColorAttributes(path, id) {
			const url = generateUrl(path + id)
			let response
			try {
				response = await axios.get(url)
			} catch (e) {
				response = e.response
			}
			this.checkForErrorCode(response.status)
			return (response.status === 200 && response.data?.color)
				? response.data.color
				: ''
		},
	},
}
</script>
<style scoped lang="scss">
.searchInput {
	width: 100%;
}

.stateMsg {
	padding: 30px;
	text-align: center;
	color: #6d6d6d;
}

</style>
