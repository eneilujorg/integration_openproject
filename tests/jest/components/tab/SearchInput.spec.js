/* jshint esversion: 8 */
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import SearchInput from '../../../../src/components/tab/SearchInput'
import workPackagesSearchResponse from '../../fixtures/workPackagesSearchResponse.json'
import workPackagesSearchResponseNoAssignee from '../../fixtures/workPackagesSearchResponseNoAssignee.json'

jest.mock('@nextcloud/l10n', () => ({
	translate: jest.fn((app, msg) => msg),
	getLanguage: jest.fn(() => ''),
}))

const localVue = createLocalVue()

function mountSearchInput() {
	return mount(SearchInput, {
		localVue,
		mocks: {
			t: (msg) => msg,
			generateUrl() {
				return '/'
			},
		},
		stubs: {
			Avatar: true,
		},
	})
}

describe('SearchInput.vue tests', () => {
	describe('state messages', () => {
		it.each([{
			state: 'no-token',
			message: 'No OpenProject account connected',
		}, {
			state: 'error',
			message: 'Error connecting to OpenProject',
		}, {
			state: 'loading',
			message: 'Wait while we fetch workpackages',
		}])('should be displayed depending upon the state', async (cases) => {
			const stateSelector = '.stateMsg'
			const wrapper = shallowMount(SearchInput, {
				localVue,
				mocks: {
					t: (msg) => msg,
					generateUrl() {
						return '/'
					},
				},
			})
			await wrapper.setData({
				state: cases.state,
			})
			expect(wrapper.find(stateSelector).exists()).toBeTruthy()
			expect(wrapper.find(stateSelector).text()).toMatch(cases.message)
		})
	})

	describe('work packages', () => {
		const searchListSelector = '.searchList'
		const inputSelector = '.multiselect__input'
		const statusSelector = '.filterProjectTypeStatus__status'
		const typeSelector = '.filterProjectTypeStatus__type'
		const assigneeSelector = '.filterAssignee'
		it('should not be displayed if the length of words in searchbar is less than or equal to three', async () => {
			const wrapper = mountSearchInput()
			const textInput = wrapper.find(inputSelector)
			await textInput.setValue('org')
			await wrapper.setData({
				searchResults: [],
			})
			expect(wrapper.find(searchListSelector).exists()).toBeFalsy()
		})

		it('should be displayed if the length of words in searchbar is more than three', async () => {
			jest.spyOn(SearchInput.methods, 'makeSearchRequest').mockImplementation()
			const wrapper = mountSearchInput()
			const textInput = wrapper.find(inputSelector)
			await textInput.setValue('organ')
			await wrapper.setData({
				searchResults: workPackagesSearchResponse,
			})
			const searchList = wrapper.find(searchListSelector)
			expect(searchList.exists()).toBeTruthy()
			expect(searchList).toMatchSnapshot()
		})

		it('should not be displayed if the length of words in searchbar decreases from more than 3 to less', async () => {
			jest.spyOn(SearchInput.methods, 'makeSearchRequest').mockImplementation()
			const wrapper = mountSearchInput()
			let textInput = wrapper.find(inputSelector)
			await textInput.setValue('orga')
			await wrapper.setData({
				searchResults: workPackagesSearchResponse,
			})
			let searchList = wrapper.find(searchListSelector)
			expect(searchList.exists()).toBeTruthy()
			textInput = wrapper.find(inputSelector)
			await textInput.setValue('org')
			await wrapper.setData({
				searchResults: [],
			})
			searchList = wrapper.find(searchListSelector)
			expect(searchList.exists()).toBeFalsy()
		})

		it('should display correct background color and text for workpackage status and type', async () => {
			jest.spyOn(SearchInput.methods, 'makeSearchRequest').mockImplementation()
			const wrapper = mountSearchInput()
			const textInput = wrapper.find(inputSelector)
			await textInput.setValue('organ')
			await wrapper.setData({
				searchResults: workPackagesSearchResponse,
			})
			const searchList = wrapper.find(searchListSelector)
			const typeCol = wrapper.find(typeSelector)
			const statusCol = wrapper.find(statusSelector)
			expect(searchList.exists()).toBeTruthy()
			expect(typeCol.element.style.color).toBe('red')
			expect(statusCol.element.style.backgroundColor).toBe('blue')

		})

		it('avatar and name should be displayed if assignee is present', async () => {
			jest.spyOn(SearchInput.methods, 'makeSearchRequest').mockImplementation()
			const wrapper = mountSearchInput()
			const textInput = wrapper.find(inputSelector)
			await textInput.setValue('organ')
			await wrapper.setData({
				searchResults: workPackagesSearchResponse,
			})
			const assignee = wrapper.find(assigneeSelector)
			expect(assignee.exists()).toBeTruthy()
			expect(assignee).toMatchSnapshot()
		})

		it('avatar and name not should be displayed if assignee is not present', async () => {
			jest.spyOn(SearchInput.methods, 'makeSearchRequest').mockImplementation()
			const wrapper = mountSearchInput()
			const textInput = wrapper.find(inputSelector)
			await textInput.setValue('organ')
			await wrapper.setData({
				searchResults: workPackagesSearchResponseNoAssignee,
			})
			const assignee = wrapper.find(assigneeSelector)
			expect(assignee.exists()).toBeFalsy()
		})
	})
})
