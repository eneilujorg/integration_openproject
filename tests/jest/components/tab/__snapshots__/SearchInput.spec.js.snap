// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput.vue tests state messages any: should display the correct state message 1`] = `
<div class="stateMsg text-center">

</div>
`;

exports[`SearchInput.vue tests state messages error: should display the correct state message 1`] = `
<div class="stateMsg text-center">
  Error connecting to OpenProject
</div>
`;

exports[`SearchInput.vue tests state messages no-token: should display the correct state message 1`] = `
<div class="stateMsg text-center">
  No OpenProject account connected
</div>
`;

exports[`SearchInput.vue tests work packages multiselect search list should display correct options list of search results 1`] = `
<div class="searchList">
  <div class="searchList__filterWorkPackage">
    <div class="filterProjectTypeStatus">
      <div class="filterProjectTypeStatus__project">
        test
      </div>
      <div class="filterProjectTypeStatus__type" style="color: red;">
        task
      </div>
      <div class="filterProjectTypeStatus__status" style="background-color: blue;">
        <div class="filterProjectTypeStatus__status__title">
          in-progress
        </div>
      </div>
    </div>
    <div class="filterAssignee">
      <div class="filterAssignee__avatar">
        <avatar-stub url="/server/index.php/apps/integration_openproject/avatar?userId=1&amp;userName=System" user="test" showuserstatus="true" showuserstatuscompact="true" displayname="test" size="25" allowplaceholder="true" menuposition="center" menucontainer="body" class="item-avatar"></avatar-stub>
      </div>
      <div class="filterAssignee__assignee">
        test
      </div>
    </div>
  </div>
  <div class="filterWorkpackageSubject">
    <div class="filterWorkpackageSubject__subject">
      Organize work-packages
    </div>
  </div>
</div>
`;
